File: sf_trading/sf_trading/report/ordered_items_pending_billing/ordered_items_pending_billing.py
# Copyright (c) 2026, Enfono Technologies and contributors
# For license information, please see license.txt

"""Sales Orders raised and not yet invoiced, by item row or by order.

The order layer of the open-item family. Its siblings all begin at a delivery or an invoice, so
an order that was placed and never billed showed up nowhere; core deleted its own version of
this report in v13 (`erpnext/patches/v13_0/delete_old_sales_reports.py`), leaving Sales Order
Analysis, which knows nothing about branch or ageing.

`Delivered Items Pending Billing` overlaps deliberately and does not contradict: it keys on the
Delivery Note, this keys on the order, and an order whose goods went out on a note appears in
both — here with its Delivered Qty column filled, so the reader can see which pending orders
have already shipped.
"""

from frappe import _

from sf_trading.open_items import (
	document_columns,
	fold_to_documents,
	ordered_items_pending_billing,
	report_columns,
	shows_documents,
)


# Where the order's own billed figure disagrees with the invoices, or something about the order
# changes what the row means — a draft invoice already raised, an advance received, a Closed
# order, a foreign currency. Said in words rather than resolved silently.
REMARKS_COLUMN = {"label": _("Remarks"), "fieldname": "remarks", "fieldtype": "Data", "width": 320}


def execute(filters=None):
	filters = filters or {}
	rows = ordered_items_pending_billing(filters)

	if shows_documents(filters):
		columns = document_columns(
			"Sales Order", "Customer", "billed_qty", _("Billed Qty"), _("Order Total")
		)
		# Delivered Qty rides along in the folded view too: it is summed by fold_to_documents
		# because the item rows carry it
		columns.insert(-2, {
			"label": _("Delivered Qty"),
			"fieldname": "delivered_qty",
			"fieldtype": "Float",
			"width": 110,
		})
		columns.append(REMARKS_COLUMN)
		return columns, fold_to_documents(rows, "Sales Order", total_field="base_grand_total")

	columns = report_columns("Sales Order", "Customer", "billed_qty", _("Billed Qty"))
	columns.insert(-2, {
		"label": _("Delivered Qty"),
		"fieldname": "delivered_qty",
		"fieldtype": "Float",
		"width": 110,
	})
	columns.append(REMARKS_COLUMN)
	return columns, rows

File: sf_trading/sf_trading/report/ordered_items_pending_billing/ordered_items_pending_billing.js
// Copyright (c) 2026, Enfono Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["Ordered Items Pending Billing"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "as_on",
			label: __("As On Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		// Optional order-date window. Left empty it reports every open order, the same way its
		// siblings report every document, so nothing that reads a total off this report moves
		// because a date was typed. As On Date still governs how much billing is counted.
		{
			fieldname: "from_date",
			label: __("From Order Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Order Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "party",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			get_query: () => {
				return { filters: { company: frappe.query_report.get_filter_value("company") } };
			},
		},
		{
			fieldname: "cost_center",
			label: __("Branch (Cost Center)"),
			fieldtype: "Link",
			options: "Cost Center",
			get_query: () => {
				return { filters: { company: frappe.query_report.get_filter_value("company") } };
			},
		},
		{
			// Item Rows or Document Rows — a Select and never a Check, because query_report.js
			// drops a filter whose value is falsy, so an unticked Check never reaches the server
			// and the view could be switched on but never off.
			fieldname: "view",
			label: __("View"),
			fieldtype: "Select",
			options: ["Item Rows", "Document Rows"].join("\n"),
			default: "Item Rows",
			reqd: 1,
		},
		{
			fieldname: "range",
			label: __("Ageing Range"),
			fieldtype: "Data",
			default: "30, 60, 90",
		},
		{
			fieldname: "overdue_after",
			label: __("Highlight Older Than (Days)"),
			fieldtype: "Int",
			default: 0,
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		const threshold = frappe.query_report.get_filter_value("overdue_after");
		if (
			threshold > 0 &&
			data &&
			data.age > threshold &&
			["age", "bucket", "pending_qty", "pending_amount"].includes(column.fieldname)
		) {
			value = `<span style="color: var(--red-600); font-weight: 600">${value}</span>`;
		}
		return value;
	},
};
